tests/accounts/test_views.py

```diff
@@ -118,33 +118,129 @@ def test_post_validate_redirects(self, monkeypatch, pyramid_request,
             name="unauthenticated_userid",
         )
 
+        route_name = "accounts.profile"
+        route_path = lambda route_name, username: "/accounts/{}/".\
+                                                  format(username)
+        # TODO: Why must I mock this request method? Why is the original
+        # route_path on the request object not already aware of routes?
+        pyramid_request.route_path = pretend.call_recorder(route_path)
+
+        username = "theuser"
         form_obj = pretend.stub(
             validate=pretend.call_recorder(lambda: True),
-            username=pretend.stub(data="theuser"),
+            username=pretend.stub(data=username),
         )
         form_class = pretend.call_recorder(lambda d, login_service: form_obj)
 
         result = views.login(pyramid_request, _form_class=form_class)
 
         assert isinstance(result, HTTPSeeOther)
-        assert result.headers["Location"] == "/"
+
+        assert result.headers["Location"] == route_path(route_name, username)
         assert result.headers["foo"] == "bar"
-        assert pyramid_request.find_service.calls == [
-            pretend.call(ILoginService, context=None),
-        ]
+
         assert form_class.calls == [
             pretend.call(pyramid_request.POST, login_service=login_service),
         ]
         assert form_obj.validate.calls == [pretend.call()]
+
         assert login_service.find_userid.calls == [pretend.call("theuser")]
+
         if with_user:
             assert new_session == {}
         else:
             assert new_session == {"a": "b", "foo": "bar"}
-        assert pyramid_request.session.invalidate.calls == [pretend.call()]
+
         assert remember.calls == [pretend.call(pyramid_request, 1)]
+        assert pyramid_request.session.invalidate.calls == [pretend.call()]
+        assert pyramid_request.find_service.calls == [
+            pretend.call(ILoginService, context=None),
+        ]
+        assert pyramid_request.route_path.calls == [
+            pretend.call(route_name, username=username),
+        ]
         assert pyramid_request.session.new_csrf_token.calls == [pretend.call()]
 
+    @pytest.mark.parametrize("with_user", [True, False])
+    def test_post_validate_no_redirects(self, monkeypatch, pyramid_request,
+                                        with_user):
+        remember = pretend.call_recorder(
```

Comment:
Much like logout, this doesn't need to duplicate all of the tests that we already have that test things like "does the login view actually login" and it should instead just focus on testing whether the login view redirects to the URL it's passed in if that url is valid, or to the default url if not. That means you can ditch most of the ``call_recorders`` and just use bare lambdas or fakes. Additionally the loop should be handled by ``parametrize`` not looping over test values inside the test case itself.